lib/queuePopulator/IngestionProducer.js

```diff
@@ -0,0 +1,210 @@
+const async = require('async');
+const bucketclient = require('bucketclient');
+const Logger = require('werelogs').Logger;
+const MetadataWrapper = require('arsenal').storage.metadata.MetadataWrapper;
+
+// const config = require('../../conf/Config');
+const constants = require('../../constants');
+const RaftLogEntry = require('../../extensions/utils/RaftLogEntry');
+
+class IngestionProducer {
+    /**
+     * Create an IngestionProducer class that helps create a snapshot of
+     * pre-existing RING backend
+     *
+     * @constructor
+     * @param {object} sourceConfig - source config with value for
+     *                                bucketdBootstrap
+     */
+    constructor(sourceConfig) {
+        this.log = new Logger('Backbeat:IngestionProducer');
+        this.ringParams = {
+            bucketdBootstrap: [sourceConfig.host],
+            bucketdLog: undefined,
+        };
+        this.ringReader = new MetadataWrapper('scality', this.ringParams,
+            bucketclient, this.log);
+        this.requestLogger = this.log.newRequestLogger();
+        this.createEntry = new RaftLogEntry();
+        this.resLog = [];
+        console.log('IngestionProducer has been created');
+    }
+
+    snapshot(raftId, done) {
+        async.waterfall([
+            next => this._getBuckets(raftId, next),
+            (bucketList, next) => this._getBucketMd(bucketList, next),
+            (bucketList, next) => this._getBucketObjects(bucketList, next),
+            (bucketList, next) =>
+                this._getBucketObjectsMetadata(bucketList, next),
+        ], err => done(err, this.resLog));
+    }
+
+    _parseBucketName(bucketKey) {
+        return bucketKey.split(constants.splitter)[1];
+    }
+
+    /**
+     * Get the list of buckets using the usersBucket
+     * Each bucket is stored as a key in the usersBucket
+     *
+     * @param {number} raftId - raft session id value
+     * @param {function} done - callback function
+     * @return {Object} list of keys that correspond to list of buckets
+     */
+    _getBuckets(raftId, done) {
+        return this.ringReader.getRaftBuckets(raftId, this.requestLogger,
+        (err, res) => {
+            if (err) {
+                this.log.error('error getting list of buckets', {
+                    method: 'IngestionProducer:getBuckets', err });
+                return done(err);
+            }
+            console.log('got buckets', res);
+            return done(null, JSON.parse(res));
+        });
+    }
+
+    /**
+     * Get the metadata for each bucket, format and send the info to kafka
+     * so that it can be ingested into MongoDB
+     *
+     * @param {object} bucketList - list of buckets
+     * @param {function} done - callback function
+     * @return {object} list of buckets
+     */
+    _getBucketMd(bucketList, done) {
+        if (bucketList.length < 1) {
+            return done(null, null);
```

Comment:
I guess you don't need to introduce the specific empty list case as `null` because it will just be processed as usual by the remaining code which is compliant with 0+ bucket list size.